teps/0147-tekton-artifacts-phase1.md

```diff
@@ -0,0 +1,285 @@
+
+
+- `pkg:generic/repo?download_url=/path/to/artifact`
+    - This path would also need to be accessible by downstream Tasks for this to work. The user of the `Task/StepAction` would need to orchestrate it carefully.
+- `pkg:generic/test-results?download_url=https://localobjectstore/filename.tar.gz`
+    - This approach would work for sharing artifacts (files/tarred-folders) between Tasks without the need for a `Persistent Volume`.
+- `pkg:github/package-url/purl-spec@244fd47e07d1004`
+- `pkg:docker/cassandra@latest`
+- `pkg:huggingface/distilbert-base-uncased@043235d6088ecd3dd5fb5ca3592b6913fd516027` (ML models)
+- `pkg:maven/org.apache.xmlgraphics/batik-anim@1.9.1`
+
+`purl-spec` also provides a common standard that `Tasks/Steps` can leverage so that output metadata information of one `Task/Step` could be consumed by another in the future (targetting phase 2). Additionally, it becomes a standard way to distinguish between types of artifacts. [package-url](https://github.com/package-url) also provides client libraries in multiple languages to parse the `purl`. This is something users could leverage.
+
+##### `digest`
+
+The `digest` component includes the result of applying an hash function to the artifact itself.
+The `digest` is in the format `hash-function:<digest>`. Example values:
+
+- `md5:471a3a9ddc1d6e699190b9f4ca88aadf`
+- `sha256:8796357729cfd877cf8fa7d45a8ab3524d9249c23a0bf68bb0026c0783b881d2`
+
+### From Step into Status
+
+The artifact provenance data (i.e. metadata content) is extracted from the local storage using the same mechanism used for results. Results and artifacts data is collated together and stored either in the termination message or in a sidecar log, depending on configuration.
+
+
+The artifact data is then read by the Tekton `TaskRun` controller and copied stored into the `Step State`:
+
+```yaml!
+status:
+  results:
+  - name: release
+    type: string
+    value: |
+      https://storage.googleapis.com/tekton-releases/pipeline/previous/v0.50.2/release.yaml
+  steps:
+    - container: step-1
+      inputs:
+        - name: source
+          value:
+            - uri: pkg:generic/source
+              digest: sha256:8796357729cfd877cf8fa7d45a8ab3524d9249c23a0bf68bb0026c0783b881d2
+  steps:
+    - container: step-2
+      outputs:
+        - name: release-file
+          value:
+            - uri: pkg:generic/release-file
+              digest: sha256:33a06c928729e52d1991a2c55765a7c30ef72b098533f220f3f1d6f352fd32e8
+```
+### From Status into Provenance
+
+The Tekton Chains formatter would access the artifact information from the `StepState`. 
+- All `inputs` will be inserted in the `resolvedDependency` section of the `SLSA provenance`.
+- All `outputs` will be placed as `subjects` 
+    - with the exception of artifacts with `uri` prefixed with `pkg:generic`.
+- All `pkg:generic` types of artifacts will be inserted as `byProducts` in the provenance.
```

Comment:
Can you dig into more detail about this? I'm skeptical that we can make the assumption that `pkg:generic` means it's a byproduct (or the reverse - that non-generic resources aren't byproducts) 🤔

Reply:
That's a good point. 

I generally assumed that `generics` would refer to things like file-system based artifacts. While things like images, packages etc. refer to more specific package types. 

If we don't want to do that then Chains cannot tell the difference here and it can simply add all `outputs` as `subjects` and `inputs` as `resolvedDependencies`. We don't add anything into `byProducts`. May be in phase2 using the Artifact API, we could revisit this distinction and further separate between subjects and byProducts via some user input. Only the `pipeline` author truly knows what is meant to be a `byProduct` vs `subject`. At the step level, that information is not known.

Reply:
+1 to not treating `pkg:generic` differently.

Reply:
Fixed.